File: lib/config/CfgAudioControl.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2025 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import "package:shared_preferences/shared_preferences.dart";

import "../constants/Strings.dart";
import "../iscp/ConnectionIf.dart";
import "../iscp/StateManager.dart";
import "../iscp/messages/EnumParameterMsg.dart";
import "../iscp/messages/ListeningModeMsg.dart";
import "../utils/Convert.dart";
import "../utils/Pair.dart";
import "../utils/Platform.dart";
import "CfgModule.dart";
import "CheckableItem.dart";

enum VolumeUnit
{
    ABSOLUTE,
    RELATIVE
}

class CfgAudioControl extends CfgModule
{
    // Sound control
    static const Pair<String, String> SOUND_CONTROL = Pair<String, String>("sound_control", Strings.pref_sound_control_default);
    String _soundControl = SOUND_CONTROL.item2;

    String get soundControl
    => _soundControl;

    set soundControl(String value)
    {
        _soundControl = value;
        saveStringParameter(SOUND_CONTROL, value);
    }

    // Force audio control
    static const Pair<String, bool> FORCE_AUDIO_CONTROL = Pair<String, bool>("force_audio_control", false);
    bool _forceAudioControl = FORCE_AUDIO_CONTROL.item2;

    bool get isForceAudioControl
    => _forceAudioControl;

    // Volume unit
    static const Pair<String, String> VOLUME_UNIT = Pair<String, String>("volume_unit", "ABSOLUTE");
    VolumeUnit _volumeUnit = VolumeUnit.ABSOLUTE;

    VolumeUnit get volumeUnit
    => _volumeUnit;

    static const Pair<String, String> ZERO_LEVEL = Pair<String, String>("zero_level", "");
    double? _zeroLevel;

    double? get zeroLevel
    => _zeroLevel;

    // Selected listening modes
    static const String _SELECTED_LISTENING_MODES = "selected_listening_modes";

    // Master volume maximum
    static const Pair<String, int> MASTER_VOLUME_MAX = Pair<String, int>("master_volume_max", 9999);
    int _masterVolumeMax = MASTER_VOLUME_MAX.item2;

    int get masterVolumeMax
    => _masterVolumeMax;

    set masterVolumeMax(int value)
    {
        _masterVolumeMax = value;
        saveIntegerParameter(getModelDependentInt(MASTER_VOLUME_MAX), value);
    }

    // Integra modes
    static const List<ListeningMode> _ISCP_LISTENING_MODES = [
        ListeningMode.MODE_0F, // MONO
        ListeningMode.MODE_00, // STEREO
        ListeningMode.MODE_01, // DIRECT
        ListeningMode.MODE_09, // UNPLUGGED
        ListeningMode.MODE_08, // ORCHESTRA
        ListeningMode.MODE_0A, // STUDIO-MIX
        ListeningMode.MODE_11, // PURE AUDIO
        ListeningMode.MODE_13, // FULL MONO
        ListeningMode.MODE_0C, // ALL CH STEREO
        ListeningMode.MODE_0B, // TV Logic
        ListeningMode.MODE_0D, // Theater-Dimensional
        ListeningMode.MODE_40, // DOLBY DIGITAL
        ListeningMode.MODE_80, // DOLBY SURROUND
        ListeningMode.MODE_84, // Dolby THX Cinema
        ListeningMode.MODE_8B, // Dolby THX Music
        ListeningMode.MODE_89, // Dolby THX Games
        ListeningMode.MODE_82, // DTS NEURAL:X
        ListeningMode.MODE_17, // DTS Virtual:X
        ListeningMode.MODE_03, // Game-RPG
        ListeningMode.MODE_05, // Game-Action
        ListeningMode.MODE_06, // Game-Rock
        ListeningMode.MODE_0E  // Game-Sports
    ];

    // Denon modes
    static const List<ListeningMode> _DCP_LISTENING_MODES = [
        ListeningMode.MODE_DCP_DIRECT,
        ListeningMode.MODE_DCP_PURE_DIRECT,
        ListeningMode.MODE_DCP_STEREO,
        ListeningMode.MODE_DCP_ALL_ZONE_STEREO,
        ListeningMode.MODE_DCP_AUTO,
        ListeningMode.MODE_DCP_DOLBY_DIGITAL,
        ListeningMode.MODE_DCP_DTS_SURROUND,
        ListeningMode.MODE_DCP_AURO3D,
        ListeningMode.MODE_DCP_AURO2DSURR,
        ListeningMode.MODE_DCP_MCH_STEREO,
        ListeningMode.MODE_DCP_WIDE_SCREEN,
        ListeningMode.MODE_DCP_SUPER_STADIUM,
        ListeningMode.MODE_DCP_ROCK_ARENA,
        ListeningMode.MODE_DCP_JAZZ_CLUB,
        ListeningMode.MODE_DCP_CLASSIC_CONCERT,
        ListeningMode.MODE_DCP_MONO_MOVIE,
        ListeningMode.MODE_DCP_MATRIX,
        ListeningMode.MODE_DCP_VIDEO_GAME,
        ListeningMode.MODE_DCP_VIRTUAL
    ];

    // Master volume hardware keys
    static const Pair<String, bool> VOLUME_KEYS = Pair<String, bool>("volume_keys", true); // For Android only
    bool _volumeKeys = VOLUME_KEYS.item2;

    bool get volumeKeys
    => _volumeKeys;

    // Zone maximum volume (used to detect external amplifier)
    static const Pair<String, String> ZONE_VOLUME_MAX = Pair<String, String>("zone_volume_max", "");
    final List<int> _zoneVolumeMax = [];

    int zoneVolumeMax(int zone)
    => zone < _zoneVolumeMax.length? _zoneVolumeMax[zone] : 0;

    // methods
    CfgAudioControl(final SharedPreferences preferences) : super(preferences);

    @override
    void read()
    {
        _soundControl = getString(SOUND_CONTROL, doLog: true);
        _forceAudioControl = getBool(FORCE_AUDIO_CONTROL, doLog: true);
        final _volumeUnitStr = getString(VOLUME_UNIT, doLog: true).toUpperCase();
        _volumeUnit = VolumeUnit.values.firstWhere(
                (p) => Convert.enumToString(p).toUpperCase() == _volumeUnitStr,
                orElse: () => VolumeUnit.ABSOLUTE);
        _zeroLevel = double.tryParse(getString(ZERO_LEVEL, doLog: true));
        _volumeKeys = Platform.isAndroid ? getBool(VOLUME_KEYS, doLog: true) : false;
        _zoneVolumeMax.clear();
        getString(ZONE_VOLUME_MAX, doLog: true).split(",").forEach((str) => _zoneVolumeMax.add(int.tryParse(str) ?? 0));
    }

    @override
    void setReceiverInformation(StateManager stateManager)
    {
        _masterVolumeMax = getInt(getModelDependentInt(MASTER_VOLUME_MAX), doLog: true);
        _zoneVolumeMax.clear();
        String volumeZoneMaxStr = "";
        stateManager.state.receiverInformation.zones.forEach((z)
        {
            _zoneVolumeMax.add(z.getVolMax);
            volumeZoneMaxStr += ("," + z.getVolMax.toString());
        });
        saveStringParameter(ZONE_VOLUME_MAX, volumeZoneMaxStr.startsWith(",") ? volumeZoneMaxStr.substring(1) : volumeZoneMaxStr);
    }

    void setVolumeUnit(VolumeUnit volumeUnit, double? zeroValue)
    {
        _volumeUnit = volumeUnit;
        saveStringParameter(VOLUME_UNIT, Convert.enumToString(_volumeUnit), prefix: "  ");
        _zeroLevel = zeroValue;
        saveStringParameter(ZERO_LEVEL, _zeroLevel == null ? "" : _zeroLevel.toString(), prefix: "  ");
    }
    
    List<EnumItem<ListeningMode>> getSortedListeningModes(
        bool allItems, EnumItem<ListeningMode> activeItem, ProtoType protoType)
    {
        final List<EnumItem<ListeningMode>> result = [];
        final List<String> defItems = [];

        getListeningModes(protoType).forEach((m)
        => defItems.add(ListeningModeMsg.ValueEnum.valueByKey(m).getCode));

        final String par = getModelDependentParameter(getSelectedListeningModePar(protoType));
        for (CheckableItem sp in CheckableItem.readFromPreference(this, par, defItems))
        {
            final bool visible = allItems || sp.checked ||
                (activeItem.key != ListeningMode.NONE && activeItem.getCode == sp.code);
            ListeningModeMsg.ValueEnum.values.forEach((m)
            {
                if (visible && ListeningModeMsg.isMode(m.key) && m.key != ListeningMode.NONE && m.getCode == sp.code)
                {
                    result.add(m);
                }
            });
        }
        return result;
    }

    static List<ListeningMode> getListeningModes(ProtoType protoType)
    => (protoType == ProtoType.ISCP) ? _ISCP_LISTENING_MODES : _DCP_LISTENING_MODES;

    static String getSelectedListeningModePar(ProtoType protoType)
    {
        String par = _SELECTED_LISTENING_MODES;
        if (protoType == ProtoType.DCP)
        {
            par += "_DCP";
        }
        return par;
    }
}

File: app/src/main/java/com/mkulesh/onpc/config/CfgAudioControl.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2022 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.config;

import android.content.Context;
import android.content.SharedPreferences;

import com.mkulesh.onpc.R;
import com.mkulesh.onpc.iscp.messages.ListeningModeMsg;
import com.mkulesh.onpc.utils.Logging;
import com.mkulesh.onpc.utils.Utils;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class CfgAudioControl
{
    static final String SOUND_CONTROL = "sound_control";
    private static final String FORCE_AUDIO_CONTROL = "force_audio_control";
    private static final String SELECTED_LISTENING_MODES = "selected_listening_modes";
    private static final String MASTER_VOLUME_MAX = "master_volume_max";
    private static final String VOLUME_KEYS = "volume_keys";

    private static final ListeningModeMsg.Mode[] ISCP_LISTENING_MODES = new ListeningModeMsg.Mode[]{
            ListeningModeMsg.Mode.MODE_0F, // MONO
            ListeningModeMsg.Mode.MODE_00, // STEREO
            ListeningModeMsg.Mode.MODE_01, // DIRECT
            ListeningModeMsg.Mode.MODE_09, // UNPLUGGED
            ListeningModeMsg.Mode.MODE_08, // ORCHESTRA
            ListeningModeMsg.Mode.MODE_0A, // STUDIO-MIX
            ListeningModeMsg.Mode.MODE_11, // PURE AUDIO
            ListeningModeMsg.Mode.MODE_0C, // ALL CH STEREO
            ListeningModeMsg.Mode.MODE_0B, // TV Logic
            ListeningModeMsg.Mode.MODE_0D, // Theater-Dimensional
            ListeningModeMsg.Mode.MODE_40, // DOLBY DIGITAL
            ListeningModeMsg.Mode.MODE_80, // DOLBY SURROUND
            ListeningModeMsg.Mode.MODE_84, // Dolby THX Cinema
            ListeningModeMsg.Mode.MODE_8B, // Dolby THX Music
            ListeningModeMsg.Mode.MODE_89, // Dolby THX Games
            ListeningModeMsg.Mode.MODE_03, // Game-RPG
            ListeningModeMsg.Mode.MODE_05, // Game-Action
            ListeningModeMsg.Mode.MODE_06, // Game-Rock
            ListeningModeMsg.Mode.MODE_0E, // Game-Sports
            ListeningModeMsg.Mode.MODE_82, // DTS NEURAL:X
            ListeningModeMsg.Mode.MODE_17  // DTS Virtual:X
    };

    private static final ListeningModeMsg.Mode[] DCP_LISTENING_MODES = new ListeningModeMsg.Mode[]{
            ListeningModeMsg.Mode.DCP_DIRECT,
            ListeningModeMsg.Mode.DCP_PURE_DIRECT,
            ListeningModeMsg.Mode.DCP_STEREO,
            ListeningModeMsg.Mode.DCP_AUTO,
            ListeningModeMsg.Mode.DCP_DOLBY_DIGITAL,
            ListeningModeMsg.Mode.DCP_DTS_SURROUND,
            ListeningModeMsg.Mode.DCP_AURO3D,
            ListeningModeMsg.Mode.DCP_AURO2DSURR,
            ListeningModeMsg.Mode.DCP_MCH_STEREO,
            ListeningModeMsg.Mode.DCP_WIDE_SCREEN,
            ListeningModeMsg.Mode.DCP_SUPER_STADIUM,
            ListeningModeMsg.Mode.DCP_ROCK_ARENA,
            ListeningModeMsg.Mode.DCP_JAZZ_CLUB,
            ListeningModeMsg.Mode.DCP_CLASSIC_CONCERT,
            ListeningModeMsg.Mode.DCP_MONO_MOVIE,
            ListeningModeMsg.Mode.DCP_MATRIX,
            ListeningModeMsg.Mode.DCP_VIDEO_GAME,
            ListeningModeMsg.Mode.DCP_VIRTUAL
    };

    private SharedPreferences preferences;
    private String soundControl;

    void setPreferences(SharedPreferences preferences)
    {
        this.preferences = preferences;
    }

    void read(final Context context)
    {
        soundControl = preferences.getString(CfgAudioControl.SOUND_CONTROL,
                context.getResources().getString(R.string.pref_sound_control_default));
    }

    public String getSoundControl()
    {
        return soundControl;
    }

    public boolean isForceAudioControl()
    {
        return preferences.getBoolean(FORCE_AUDIO_CONTROL, false);
    }

    @NonNull
    private String getMasterVolumeMaxParameter()
    {
        return MASTER_VOLUME_MAX + "_" + preferences.getString(Configuration.MODEL, "NONE");
    }

    public int getMasterVolumeMax()
    {
        return preferences.getInt(getMasterVolumeMaxParameter(), Integer.MAX_VALUE);
    }

    public void setMasterVolumeMax(int limit)
    {
        Logging.info(this, "Save volume max limit: " + limit);
        SharedPreferences.Editor prefEditor = preferences.edit();
        prefEditor.putInt(getMasterVolumeMaxParameter(), limit);
        prefEditor.apply();
    }

    @NonNull
    public ArrayList<ListeningModeMsg.Mode> getSortedListeningModes(
            boolean allItems,
            @Nullable ListeningModeMsg.Mode activeItem,
            @NonNull Utils.ProtoType protoType)
    {
        final ArrayList<ListeningModeMsg.Mode> result = new ArrayList<>();
        final ArrayList<String> defItems = new ArrayList<>();
        for (ListeningModeMsg.Mode i : getListeningModes(protoType))
        {
            defItems.add(i.getCode());
        }
        final String par = getSelectedListeningModePar(protoType);
        for (CheckableItem sp : CheckableItem.readFromPreference(preferences, par, defItems))
        {
            final boolean visible = allItems || sp.checked ||
                    (activeItem != null && activeItem.getCode().equals(sp.code));
            for (ListeningModeMsg.Mode i : getListeningModes(protoType))
            {
                if (visible && i.getCode().equals(sp.code))
                {
                    result.add(i);
                }
            }
        }
        return result;
    }

    public boolean isVolumeKeys()
    {
        return preferences.getBoolean(VOLUME_KEYS, false);
    }

    public static ListeningModeMsg.Mode[] getListeningModes(@NonNull Utils.ProtoType protoType)
    {
        return protoType == Utils.ProtoType.ISCP ? ISCP_LISTENING_MODES : DCP_LISTENING_MODES;
    }

    public static String getSelectedListeningModePar(@NonNull Utils.ProtoType protoType)
    {
        String par = SELECTED_LISTENING_MODES;
        if (protoType == Utils.ProtoType.DCP)
        {
            par += "_DCP";
        }
        return par;
    }
}
